File: src/main/java/de/maggiwuerze/xdccloader/controller/ChannelController.java
package de.maggiwuerze.xdccloader.controller;

import de.maggiwuerze.xdccloader.model.entity.Bot;
import de.maggiwuerze.xdccloader.model.entity.Channel;
import de.maggiwuerze.xdccloader.model.forms.ChannelForm;
import de.maggiwuerze.xdccloader.model.transport.ChannelTO;
import de.maggiwuerze.xdccloader.service.ChannelService;
import java.security.Principal;
import java.util.List;
import java.util.stream.Collectors;
import javax.validation.ConstraintViolationException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

@Slf4j
@RequiredArgsConstructor
@RestController
class ChannelController {

	private final ChannelService channelService;

	/**
	 * @return a list of all channels
	 */
	@GetMapping("/channels/")
	public ResponseEntity<List<ChannelTO>> getAllChannels() {
		List<ChannelTO> channels = channelService.list().stream().map(ChannelTO::new).collect(Collectors.toList());
		return new ResponseEntity<>(channels, HttpStatus.OK);
	}

	@PostMapping("/channels/")
	public ResponseEntity<?> addChannel(@RequestBody ChannelForm channelForm) {
		try {
			Channel channel = channelService.save(new Channel(channelForm.getName()));
			return new ResponseEntity<>("Download added successfully. id=[" + channel.getId() + "]", HttpStatus.OK);
		} catch (ConstraintViolationException e) {
			log.error(e.getMessage());
			return new ResponseEntity<>(e.getConstraintViolations().stream().findFirst().get().getMessage(), HttpStatus.BAD_REQUEST);
		}
	}

	/**
	 * @return an example channel object to populate the attributes for channel creation popover
	 */
	@GetMapping("/channels/example")
	public ResponseEntity<List<Channel>> getExampleChannel(Principal principal) {
		return new ResponseEntity<>(List.of(new Channel()), HttpStatus.OK);
	}

	@DeleteMapping("/channels/")
	public ResponseEntity<?> delete( Long channelId) {
		try {
			channelService.delete(channelId);
			return new ResponseEntity<>("Channel deleted successfully.", HttpStatus.OK);
		}catch (Exception e) {
			return new ResponseEntity<>("Channel could not be deleted", HttpStatus.CONFLICT);
		}

	}

}

File: src/main/kotlin/de/maggiwuerze/xdccwebloader/controller/ChannelController.kt
package de.maggiwuerze.xdccwebloader.controller

import de.maggiwuerze.xdccwebloader.model.entity.Channel
import de.maggiwuerze.xdccwebloader.model.entity.ChannelTO
import de.maggiwuerze.xdccwebloader.model.forms.ChannelFormTO
import de.maggiwuerze.xdccwebloader.service.ChannelService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.*

@RestController
@RequestMapping("channels/")
class ChannelController(val channelService: ChannelService) {

    /**
     * @return a list of all channels
     */
    @GetMapping()
    fun listChannels(): ResponseEntity<List<ChannelTO>> =
        ResponseEntity(channelService.list().map { it.toTO() }, HttpStatus.OK)

    @GetMapping("{id}")
    fun getChannel(@PathVariable id: UUID): ResponseEntity<ChannelTO> =
        channelService.findById(id)?.let { ResponseEntity(it.toTO(), HttpStatus.OK) }
            ?: ResponseEntity(HttpStatus.NOT_FOUND)

    @PutMapping("{id}")
    fun updateChannel(@PathVariable id: UUID, @RequestBody channelFormTO: ChannelFormTO): ResponseEntity<ChannelTO> =
        ResponseEntity(channelService.update(id, channelFormTO).toTO(), HttpStatus.OK)

    @DeleteMapping("{id}")
    fun deleteChannel(@PathVariable id: UUID): ResponseEntity<HttpStatus> =
        ResponseEntity(channelService.delete(id))

    @PostMapping
    fun createChannel(@RequestBody channelFormTO: ChannelFormTO): ResponseEntity<ChannelTO> {
        channelService.save(Channel(name = channelFormTO.name)).let {
            return ResponseEntity(it.toTO(), HttpStatus.OK)
        }
    }


}